File: javascript/Phylo/Mediators/TaxaMediator.js
Phylo.Mediators = {
	'TaxaMediator' : {
		'object'       : [],
		'relationship' : []
	}
};

Phylo.Mediators.TaxaMediator.register = function(obj) {
    var id = obj.get_id();
    Phylo.Mediators.TaxaMediator.object[id] = obj;
    return this;
};

Phylo.Mediators.TaxaMediator.unregister = function(obj) {
    var id = obj.get_id();
    if ( Phylo.Mediators.TaxaMediator.object[id] != null ) {
        // one-to-many relationship
        if ( Phylo.Mediators.TaxaMediator.relationship[id] != null ) {
            delete Phylo.Mediators.TaxaMediator.relationship[id];
        }
        // one-to-one relationship
        else {
            for ( var i = 0; i < Phylo.Mediators.TaxaMediator.relationship.length; i++ ) {
                var relation = Phylo.Mediators.TaxaMediator.relationship[i];
                if ( relation[id] != null ) {
                    delete relation[id];
                    break;
                }
            }
        }
        delete Phylo.Mediators.TaxaMediator.object[id];
    }
    return this;
};

Phylo.Mediators.TaxaMediator.set_link = function(args) {
    var one  = args["one"];
    var many = args["many"];
    var one_id  = one.get_id();
    var many_id = many.get_id();
    for ( var i = 0; i < Phylo.Mediators.TaxaMediator.relationship.length; i++ ) {
        if ( Phylo.Mediators.TaxaMediator.relationship[i] != null && Phylo.Mediators.TaxaMediator.relationship[i][many_id.toString()] != null ) {
            delete Phylo.Mediators.TaxaMediator.relationship[i][many_id];
            break;
        }
    }
    if ( Phylo.Mediators.TaxaMediator.relationship[one_id] == null ) {
        Phylo.Mediators.TaxaMediator.relationship[one_id] = {};
    }
    Phylo.Mediators.TaxaMediator.relationship[one_id][many.get_id()] = many._type();
    return this;
};

Phylo.Mediators.TaxaMediator.get_link = function (args) {
    var id = args["source"].get_id();
    
    // have to get many objects of the same type
    if ( args["type"] != null ) {
        if ( Phylo.Mediators.TaxaMediator.relationship[id] == null ) {
            return null;
        }
        var result = new Array();
        for ( var key in Phylo.Mediators.TaxaMediator.relationship[id] ) {
            if ( Phylo.Mediators.TaxaMediator.relationship[id][key] == args["type"] ) {
                result.push(Phylo.Mediators.TaxaMediator.object[key]);
            }
        }
        return result;
    }
    
    // have to get just one
    else {
        for ( var i = 0; i < Phylo.Mediators.TaxaMediator.relationship.length; i++ ) {
            if ( Phylo.Mediators.TaxaMediator.relationship[i] != null && Phylo.Mediators.TaxaMediator.relationship[i][id] != null ) {
                return Phylo.Mediators.TaxaMediator.object[i];
            }
        }
    }
    return null;
};

Phylo.Mediators.TaxaMediator.remove_link = function(args) {
    var one = args["one"];
    var many = args["many"];
    if ( one != null ) {
        var one_id = one.get_id();
        var relation = Phylo.Mediators.TaxaMediator.relationship[one_id];
        if ( relation == null ) {
            return this;
        }
        var many_id = many.get_id();
        delete relation[many_id];
        return this;
    }
    else {
        var id = many.get_id();
        for ( var i = 0; i < Phylo.Mediators.TaxaMediator.relationship.length; i++ ) {
            var found_relation = Phylo.Mediators.TaxaMediator.relationship[i];
            if ( found_relation[id] != null ) {
                delete found_relation[id];
                break;
            }
        }
        return this;
    }
};

File: java/src/org/biophylo/mediators/TaxaMediator.java
package org.biophylo.mediators;
import java.util.*;
import org.biophylo.taxa.*;
import org.biophylo.util.*;
import org.biophylo.*;

public class TaxaMediator {
	private static TaxaMediator instance = null;
	private static Logger logger = Logger.getInstance();
	private ObjectMediator object;
	private HashMap objectsForTaxon;
	private HashMap taxonForObject;
	
	/**
	 * 
	 */
	protected TaxaMediator() {
	     this.object = ObjectMediator.getInstance();
	     this.objectsForTaxon = new HashMap(); // key is taxon ID
	     this.taxonForObject = new HashMap(); // key is taxonlinker id
	}
	
	/**
	 * @return
	 */
	public static TaxaMediator getInstance() {
		if(instance == null) {
			instance = new TaxaMediator();
		}
	    return instance;
	}
	
	/**
	 * @param taxonId
	 * @param linkerId
	 */
	public void setLink(int taxonId,int linkerId) {
		Integer lId = new Integer(linkerId);
		Integer tId = new Integer(taxonId);
		Containable obj = (Containable)this.object.getObjectById(linkerId);
		Integer type = new Integer(obj.type());
		this.taxonForObject.put(lId, tId);
		if ( ! this.objectsForTaxon.containsKey(tId) ) {
			this.objectsForTaxon.put(tId, new HashMap());
		}
		((HashMap)this.objectsForTaxon.get(tId)).put(lId, type);
	}
	
	/**
	 * @param linkerId
	 * @return
	 */
	public Containable getLink(int linkerId) {
		Integer lId = new Integer(linkerId);
		if ( this.taxonForObject.containsKey(lId) ) {
			Integer taxonId = (Integer)this.taxonForObject.get(lId);
			return (Containable)this.object.getObjectById(taxonId.intValue());
		}
		return null;
	}
	
	/**
	 * @param taxonId
	 * @param type
	 * @return
	 */
	public Vector getLink(int taxonId, int type) {
		Integer tId = new Integer(taxonId);
		if ( this.objectsForTaxon.containsKey(tId) ) {
			logger.debug("contains taxon id " + taxonId + " (" + this.object.getObjectById(taxonId) + ")");
			HashMap objects = (HashMap)this.objectsForTaxon.get(tId);
			logger.debug("HashMap: " + objects);
			Integer objType = new Integer(type);
			logger.debug("Requested object type " + type);
			if ( objects.containsValue(objType) ) {
				logger.debug("Have link to that object type");
				Object keys[] = objects.keySet().toArray();
				Vector result = new Vector();
				for ( int i = 0; i < keys.length; i++ ) {					
					Integer key = (Integer)keys[i];
					logger.debug("Inspecting obj "+keys[i]+ " => type "+((Integer)objects.get(key)).intValue());
					if ( ((Integer)objects.get(key)).intValue() == type ) {
						logger.debug("Found object id "+key.toString());
						result.add(this.object.getObjectById(key.intValue()));
					}
				}
				return result;
			}

		}
		return null;
	}
	
	/**
	 * @param taxonId
	 * @param linkerId
	 */
	public void removeLink(int taxonId, int linkerId) {
		logger.info("removing link " + taxonId + " => " + linkerId);
		Integer tId = new Integer(taxonId);
		Integer lId = new Integer(linkerId);
		if ( this.objectsForTaxon.containsKey(tId) ) {
			if ( ((HashMap)this.objectsForTaxon.get(tId)).containsKey(lId) ) {
				((HashMap)this.objectsForTaxon.get(tId)).remove(lId);
			}
			else {
				this.objectsForTaxon.remove(tId);
			}
		}
		if ( this.taxonForObject.containsKey(lId) ) {
			this.taxonForObject.remove(lId);
		}
	}
}
